Lior core concepts revision 2 (#3237)
enkidevs/curriculum: coding-intro/coding-intro-core/core-concepts/programs-and-data.md

---
author: kapnobatai136
tags:
  - discussion
type: normal

> category: must-know

Reviewer: same here, "category", not "tags"

---
## Content

Let's introduce the building blocks we'll cover in this course.

### Programs

Just like you start your day with a to-do list, we give computers their own to-do lists, known as programs, filled with very specific tasks to complete.

> 💡 Scroll down to the Learn more section to see a hilarious example of how specific instructions computers need 😆

### Data

Data is the information that our programs work with. In code, we store these in variables.

For instance, we could store someone's `name` and `age` in variables[1]:
```python
name = "Batman"
age = 30
```

### Types

It’s easy to divide 24 by 2 and get 12. However, dividing "Batman" by 2 might land you in jail.

That's why we group our data into distinct **data types**, the three main ones being:

| Type    | Examples    |
| ------- | ----------- |
| String  | "Batman"    |
| Number  | 5,-0.5     |
| Boolean | True, False |

These data types save us from attempting number-related operations on texts and vice versa.

> 🐈 Can a computer tell us what's the square root of "cat"?
>
> Leave a comment or view some of the other comments for inspiration before moving on.

## Footnotes

[1: What language is this?]

> Every language has its own slightly different way of creating variables. The examples in this insight are all in python

Reviewer: ```suggestion
Every language has its own slightly different way of creating variables. The examples in this insight are all in python.
```

---

## Practice

In the code below, raining is a ???, temperature is a ???, and forecast is a ???.

> ```python

Reviewer: ```suggestion

```python
```

links:
  - >-
    [Exact Instructions Challenge](https://youtube.com/watch?v=Ct-lOOUqmyY){video}
practiceQuestion:
  formats:
    - fill-in-the-gap
  context: standalone

---

# Programs & Data

---
## Content

Let's introduce the building blocks we'll cover in this course.

### Programs

> Today you might have woken up with a to-do list, which is essentially your program for the day.
> 
> Similarly, we give the computers a to-do list of extremely specific commands, together called a program, to carry out.

Reviewer: Can probably simplify these into a single sentence.

Some ideas:

```suggestion
Just like you wake up with a daily to-do list, we give computers their own super-detailed to-do list, which we call a program.
```

or

```suggestion
Just like how you follow your to-do list for the day, computers follow their own list of specific tasks, known as a program.
```

or

```suggestion
Just like you start your day with a to-do list, we give computers their own to-do lists, known as programs, filled with very specific tasks to complete.
```

> 💡 Scroll down to the Learn more section to see a hilarious example of how specific instructions computers need 😆

### Data

Data is the information that our programs work with. In code, we store these in variables.

For instance, we could store someone's `name` and `age` in variables[1]:
```python
name = "Batman"
age = 30
```

### Types

> Now remember, computers are extremely stupid machines! If you write a program to divide "Batman" by 2, it is going to try (and fail) to do that.

Reviewer: Instead of "stupid", maybe say "dumb" or "annoyingly literal" or "blindly literal"?

> Now remember, computers are extremely literal machines! If you write a program to divide "Batman" by 2, it is going to try (and fail) to do that.

Reviewer: This sentence is good but not sure it's an improvement.

> It’s easy to divide 24 by 2 and get 12. However, dividing “Batman” by 2 might land you in jail.

I'd consider leaving this playful copy in there somehow. IIRC, it generated some fun engagement.

Lets dig into the comments on this insight and see what people say?

That's why we group our data into distinct **data types**, the three main ones being:

| Type    | Examples    |
| ------- | ----------- |
| String  | "Batman"    |
| Number  | 5,-0.5     |
| Boolean | True, False |

> What would we get if we tried to take the square root of `"cat"`?

Reviewer: We never explained what happens when you perform operations on a type that aren't appropriate for that type. It might be confusing to beginners how to answer this.

Before we explained:

> These data types save us from attempting number related operations on texts and vice versa.
>
> For instance, what's the square root of `"cat"`?

Note: Since this insight ends in a question, lets also change the category of this insight into "discussion".

---
## Footnotes

[1: What language is this?]
Every language has its own slightly different way of creating variables. The examples in this insight are all in python

---

## Practice

In the code below, raining is a ???, temperature is a ???, and forecast is a ???.
```python
raining = True
temperature = 42
forecast = "Today it is 42 degrees and raining"
```

- Boolean
- Number

> - String

Reviewer: 👍

Reviewer: Let's add another question for revision?

### Types

> It’s easy to divide 24 by 2 and get 12. However, dividing “Batman” by 2 might land you in jail.

Reviewer: BTW, make sure to change all of iOS quotes to standardized ones across the PR:

```suggestion
It’s easy to divide 24 by 2 and get 12. However, dividing "Batman" by 2 might land you in jail.
```

| Number  | 5,-0.5     |
| Boolean | True, False |

These data types save us from attempting number-related operations on texts and vice versa.

> > <emoji> <question>
> >
> > Leave a comment or view some of the other comments for inspiration before moving on.

Reviewer: What's the emoji and question here?

I left <emoji> and <question> as placeholders in my comment 😅 

Based on your latest changes from above, we can put back the square root of "cat" again:

```suggestion
> 🐈 Can a computer tell us what's the square root of "cat"?
>
> Leave a comment or view some of the other comments for inspiration before moving on.
```

- String

---

## Revision

> Can we divide a string by a number?
> ???

Reviewer: ```suggestion
Can we divide a string by a number?

???
```
